fuel-vm/src/interpreter/alu/wideint.rs

```diff
@@ -0,0 +1,365 @@
+use ethnum::U256;
+
+use fuel_asm::{wideint::*, PanicReason};
+use fuel_types::{RegisterId, Word};
+
+use super::super::{internal::inc_pc, is_unsafe_math, is_wrapping, ExecutableTransaction, Interpreter};
+use crate::interpreter::memory::{read_bytes, write_bytes};
+use crate::{constraints::reg_key::*, error::RuntimeError};
+
+// This macro is used to duplicate the implementation for both 128-bit and 256-bit versions.
+// It takes two type parameters: the current type and type that has double-width of it.
+// The appropriate type is chosen based on benchmarks for each operation.
+// Currently, `primitive_types` is used for anything requiring division, modulo or 512-bit precision.
+// Otherwise, `ethnum` is used for 256-bit operations, and the builtin `u128` for 128-bit operations.
+macro_rules! wideint_ops {
+    ($t:ident, $wider_t:ident) => {
+        paste::paste! {
+            // Conversion helpers
+
+            /// Converts to `primitive_types` version of the current type
+            fn [<to_prim_ $t:lower>](value: $t) -> primitive_types::[<$t:upper>] {
+                let mut buffer = [0u8; core::mem::size_of::<$t>()];
+                buffer[..].copy_from_slice(&value.to_le_bytes());
+                primitive_types::[<$t:upper>]::from_little_endian(&buffer)
+            }
+
+            /// Converts to `primitive_types` version that has double the size of the current type
+            fn [<to_wider_prim_ $t:lower>](value: $t) -> primitive_types::[<$wider_t:upper>] {
+                let mut buffer = [0u8; 2 * core::mem::size_of::<$t>()];
+                buffer[..core::mem::size_of::<$t>()].copy_from_slice(&value.to_le_bytes());
+                primitive_types::[<$wider_t:upper>]::from_little_endian(&buffer)
+            }
+
+            /// Converts to `u128` or `ethnum::U256`
+            fn [<from_prim_ $t:lower>](value: primitive_types::[<$t:upper>]) -> $t {
+                let mut buffer = [0u8; ::core::mem::size_of::<$t>()];
+                value.to_little_endian(&mut buffer);
+                $t::from_le_bytes(buffer)
+            }
+
+            /// Drops higher half of the value and converts to `u128` or `ethnum::U256`
+            fn [<truncate_from_prim_ $t:lower>](value: primitive_types::[<$wider_t:upper>]) -> $t {
+                const S: usize = ::core::mem::size_of::<$t>();
+                let mut buffer = [0u8; 2 * S];
+                value.to_little_endian(&mut buffer);
+                let truncated: [u8; S] = buffer[..S].try_into().unwrap_or_else(|_| unreachable!());
+                $t::from_le_bytes(truncated)
+            }
+
+            impl<S, Tx> Interpreter<S, Tx>
```

Comment:
should we be concerned with macro hygiene here?

Reply:
Please clarify. I'm not sure what the potential issue would be, as the macros are expanded below in the same file.